### PR TITLE
AIO read & write implemented.

### PR DESCRIPTION
**Pull request by janlemcke**
AIO write & read implemented.

### CODE REVIEW

File: src/benchmark/file_io_read_micro_benchmark.cpp [hunk 1]
```diff
@@ -13,6 +14,16 @@
...
+void aio_read_error_handling(aiocb* aiocb, uint32_t EXPECTED_BYTES){
+  const auto err = aio_error(aiocb);
+  const auto ret = aio_return(aiocb);
+
+  Assert(err == 0, "Error at aio_error(): " + strerror(errno));
+
+  Assert(ret == static_cast<int32_t>(EXPECTED_BYTES), "Error at aio_return(). Got: " + std::to_string(ret) +
+                                                          " Expected: " + std::to_string(EXPECTED_BYTES) + ".");
+}
+
```
Comment (SanJSp) [inline src/benchmark/file_io_read_micro_benchmark.cpp:26]: Please move this to the utils

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 2]
```diff
@@ -22,6 +23,39 @@ void write_data_using_pwrite(const size_t from, const size_t to, int32_t fd, uin
...
 
+void aio_error_handling(aiocb* aiocb, uint32_t expected_bytes) {
+  const auto err = aio_error(aiocb);
+  const auto ret = aio_return(aiocb);
+
+  Assert(err == 0, "Error at aio_error(): " + std::strerror(errno));
+
+  Assert(ret == static_cast<int32_t>(expected_bytes),
+         "Error at aio_return(). Got: " + std::to_string(ret) + " Expected: " + std::to_string(expected_bytes) + ".");
+}
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:34]: should be redundant when moved to utils

File: src/benchmark/file_io_read_micro_benchmark.cpp [hunk 3]
```diff
@@ -56,6 +68,51 @@ void read_data_randomly_using_pread(const size_t from, const size_t to, int32_t
...
+  struct aiocb aiocb;
+
+  memset(&aiocb, 0, sizeof(struct aiocb));
+  aiocb.aio_fildes = fd;
+  aiocb.aio_offset = from * uint32_t_size;
+  aiocb.aio_buf = read_data_start + from;
+  aiocb.aio_nbytes = bytes_to_read;
+  aiocb.aio_lio_opcode = LIO_READ;
+
+  Assert(aio_read(&aiocb) != -1, "Read error: " + strerror(errno));
```
Comment (SanJSp) [inline src/benchmark/file_io_read_micro_benchmark.cpp:83]: Can you maybe set a constant for -1? Currently -1 is a magic number

File: src/benchmark/file_io_read_micro_benchmark.cpp [hunk 4]
```diff
@@ -355,6 +408,173 @@ void FileIOMicroReadBenchmarkFixture::pread_atomic_random_multi_threaded(benchma
...
+    memset(&aiocb, 0, sizeof(struct aiocb));
+    aiocb.aio_fildes = fd;
+    aiocb.aio_offset = 0;
+    aiocb.aio_buf = std::data(read_data);
+    aiocb.aio_nbytes = NUMBER_OF_BYTES;
+    aiocb.aio_lio_opcode = LIO_READ;
+
+    Assert(aio_read(&aiocb) != -1, "Read error: " + strerror(errno));
+
+    int err;
```
Comment (SanJSp) [inline src/benchmark/file_io_read_micro_benchmark.cpp:434]: auto

File: src/benchmark/file_io_read_micro_benchmark.cpp [hunk 5]
```diff
@@ -355,6 +408,173 @@ void FileIOMicroReadBenchmarkFixture::pread_atomic_random_multi_threaded(benchma
...
+
+  for (auto _ : state) {
+    state.PauseTiming();
+
+    micro_benchmark_clear_disk_cache();
+    auto read_data = std::vector<uint32_t>{};
+    read_data.resize(NUMBER_OF_ELEMENTS);
+    state.ResumeTiming();
+
+    struct aiocb aiocb;
```
Comment (SanJSp) [inline src/benchmark/file_io_read_micro_benchmark.cpp:423]: can we initialize this using auto?

File: src/benchmark/file_io_read_micro_benchmark.cpp [hunk 6]
```diff
@@ -355,6 +408,173 @@ void FileIOMicroReadBenchmarkFixture::pread_atomic_random_multi_threaded(benchma
...
+    micro_benchmark_clear_disk_cache();
+    const auto random_indices = generate_random_indexes(NUMBER_OF_ELEMENTS);
+    auto read_data = std::vector<uint32_t>{};
+    read_data.resize(NUMBER_OF_ELEMENTS);
+    auto* read_data_start = std::data(read_data);
+
+
+    state.ResumeTiming();
+
+    struct aiocb aiocb;
```
Comment (SanJSp) [inline src/benchmark/file_io_read_micro_benchmark.cpp:505]: auto?

File: src/benchmark/file_io_read_micro_benchmark.cpp [hunk 7]
```diff
@@ -355,6 +408,173 @@ void FileIOMicroReadBenchmarkFixture::pread_atomic_random_multi_threaded(benchma
...
+    aiocb.aio_nbytes = uint32_t_size;
+
+    // TODO(everyone) Randomize inidzes to not read all the data but really randomize
+    for (auto index = size_t{0}; index < NUMBER_OF_ELEMENTS; ++index) {
+      aiocb.aio_offset = uint32_t_size * random_indices[index];
+      aiocb.aio_buf = read_data_start + index;
+
+      Assert(aio_read(&aiocb) != -1, "Read error: " + strerror(errno));
+
+      int err;
```
Comment (SanJSp) [inline src/benchmark/file_io_read_micro_benchmark.cpp:519]: auto, please specify also what int size

File: src/benchmark/file_io_read_micro_benchmark.cpp [hunk 8]
```diff
@@ -435,18 +673,25 @@ BENCHMARK_DEFINE_F(FileIOMicroReadBenchmarkFixture, IN_MEMORY_READ_RANDOM)(bench
...
     ->UseRealTime();
 BENCHMARK_REGISTER_F(FileIOMicroReadBenchmarkFixture, PREAD_ATOMIC_RANDOM_THREADED)
-    ->ArgsProduct({{10, 100, 1000}, {1, 2, 4, 6, 8, 16, 32}})
+    ->ArgsProduct({{10, 100, 1000}, {1, 2, 4, 6, 8, 16, 24, 32, 48}})
+    ->UseRealTime();
+BENCHMARK_REGISTER_F(FileIOMicroReadBenchmarkFixture, AIO_SEQUENTIAL_THREADED)
+    ->ArgsProduct({{10, 100, 1000}, {1, 2, 4, 6, 8, 16, 24, 32, 48}})
+    ->UseRealTime();
+BENCHMARK_REGISTER_F(FileIOMicroReadBenchmarkFixture, AIO_RANDOM_THREADED)
+    ->ArgsProduct({{10, 100, 1000}, {1, 2, 4, 6, 8, 16, 24, 32, 48}})
```
Comment (SanJSp) [inline src/benchmark/file_io_read_micro_benchmark.cpp:694]: great idea! Could you then also change all write benchmarks? I'm not sure if write mmap is included

File: src/benchmark/file_io_read_mmap_benchmark.cpp [hunk 9]
```diff
@@ -155,17 +155,17 @@ BENCHMARK_DEFINE_F(FileIOMicroReadBenchmarkFixture, MMAP_ATOMIC_MAP_SHARED_SEQUE
...
+    ->ArgsProduct({{100}, {1, 2, 4, 8, 16, 24, 32, 48}})
     ->UseRealTime();
 
 BENCHMARK_REGISTER_F(FileIOMicroReadBenchmarkFixture, MMAP_ATOMIC_MAP_SHARED_SEQUENTIAL)
-    ->ArgsProduct({{100}, {1, 2, 4, 8, 16, 32}})
+    ->ArgsProduct({{100}, {1, 2, 4, 8, 16, 24, 32, 48}})
     ->UseRealTime();
 BENCHMARK_REGISTER_F(FileIOMicroReadBenchmarkFixture, MMAP_ATOMIC_MAP_SHARED_RANDOM)
-    ->ArgsProduct({{100}, {1, 2, 4, 8, 16, 32}})
+    ->ArgsProduct({{100}, {1, 2, 4, 8, 16, 24, 32, 48}})
```
Comment (SanJSp) [inline src/benchmark/file_io_read_mmap_benchmark.cpp:168]: IÄm not sure why filesize here is only 100, but could you change it back to 1, 10 and 100?

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 10]
```diff
@@ -22,6 +23,39 @@ void write_data_using_pwrite(const size_t from, const size_t to, int32_t fd, uin
...
+
+void write_data_using_aio(const size_t from, const size_t to, int32_t fd, uint32_t* data_to_write_start) {
+  const auto uint32_t_size = ssize_t{sizeof(uint32_t)};
+  const auto bytes_to_write = static_cast<ssize_t>(uint32_t_size * (to - from));
+
+  struct aiocb aiocb;
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:40]: auto?

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 11]
```diff
@@ -22,6 +23,39 @@ void write_data_using_pwrite(const size_t from, const size_t to, int32_t fd, uin
...
+  memset(&aiocb, 0, sizeof(struct aiocb));
+  aiocb.aio_fildes = fd;
+  aiocb.aio_buf = data_to_write_start + from;
+  aiocb.aio_offset = from * uint32_t_size;
+  aiocb.aio_nbytes = bytes_to_write;
+  aiocb.aio_lio_opcode = LIO_WRITE;
+
+  Assert(aio_write(&aiocb) != -1, "Read error: " + std::strerror(errno));
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:48]: -1 magic number

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 12]
```diff
@@ -150,6 +184,92 @@ void FileIOWriteMicroBenchmarkFixture::pwrite_atomic_multi_threaded(benchmark::S
   }
 }
 
+void FileIOWriteMicroBenchmarkFixture::aio_single_threaded(benchmark::State& state) {
+  auto fd = int32_t{};
+  if ((fd = open(filename, O_WRONLY)) < 0) {
+    close(fd);
+    Fail("Open error:" + std::strerror(errno));
+  }
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:192]: Assert

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 13]
```diff
@@ -150,6 +184,92 @@ void FileIOWriteMicroBenchmarkFixture::pwrite_atomic_multi_threaded(benchmark::S
...
+    // The standard aio control block. A structure, that holds the information about the asnyc IO op.
+    struct aiocb aiocb;
+    memset(&aiocb, 0, sizeof(struct aiocb));
+    aiocb.aio_fildes = fd;
+    aiocb.aio_offset = 0;
+    aiocb.aio_buf = data_to_write_start;
+    aiocb.aio_nbytes = NUMBER_OF_BYTES;
+    aiocb.aio_lio_opcode = LIO_WRITE;
+
+    Assert(aio_write(&aiocb) != -1, "Read error: " + std::strerror(errno));
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:210]: -1 magic num

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 14]
```diff
@@ -150,6 +184,92 @@ void FileIOWriteMicroBenchmarkFixture::pwrite_atomic_multi_threaded(benchmark::S
...
+    sanity_check();
+    state.ResumeTiming();
+  }
+
+  close(fd);
+}
+
+void FileIOWriteMicroBenchmarkFixture::aio_multi_threaded(benchmark::State& state, uint16_t thread_count) {
+  auto filedescriptors = std::vector<int32_t>(thread_count);
+  for (auto i = size_t{0}; i < thread_count; i++) {
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:230]:
```suggestion
  for (auto index = size_t{0}; index < thread_count; ++index) {
```

please change also in line 236

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 15]
```diff
@@ -150,6 +184,92 @@ void FileIOWriteMicroBenchmarkFixture::pwrite_atomic_multi_threaded(benchmark::S
...
+    auto fd = int32_t{};
+    if ((fd = open(filename, O_WRONLY)) < 0) {
+      close(fd);
+      Fail("Open error:" + std::strerror(errno));
+    }
+    filedescriptors[i] = fd;
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:236]:
```suggestion
    filedescriptors[index] = fd;
```

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 16]
```diff
@@ -150,6 +184,92 @@ void FileIOWriteMicroBenchmarkFixture::pwrite_atomic_multi_threaded(benchmark::S
...
+  auto batch_size = static_cast<uint64_t>(std::ceil(static_cast<float>(NUMBER_OF_ELEMENTS) / thread_count));
+
+  for (auto _ : state) {
+    state.PauseTiming();
+    micro_benchmark_clear_disk_cache();
+    state.ResumeTiming();
+
+    auto* data_to_write_start = std::data(data_to_write);
+
+    for (auto i = size_t{0}; i < thread_count; i++) {
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:249]: use index instead of i

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 17]
```diff
@@ -150,6 +184,92 @@ void FileIOWriteMicroBenchmarkFixture::pwrite_atomic_multi_threaded(benchmark::S
...
+      auto from = batch_size * i;
+      auto to = from + batch_size;
+      if (to >= NUMBER_OF_ELEMENTS) {
+        to = NUMBER_OF_ELEMENTS;
+      }
+      threads[i] = std::thread(write_data_using_aio, from, to, filedescriptors[i], data_to_write_start);
+    }
+
+    for (auto i = size_t{0}; i < thread_count; i++) {
```
Comment (SanJSp) [inline src/benchmark/file_io_write_micro_benchmark.cpp:258]: index

**Review by SanJSp — CHANGES_REQUESTED** [review]
Finally async! Thanks for your efforts!

Only minor things to change :)

File: src/benchmark/file_io_read_micro_benchmark.cpp (hunk 5 above)
Comment (janlemcke) [reply src/benchmark/file_io_read_micro_benchmark.cpp:423]: as far as i read, it's not possible.

File: src/benchmark/file_io_write_micro_benchmark.cpp (hunk 14 above)
Comment (janlemcke) [reply src/benchmark/file_io_write_micro_benchmark.cpp:230]: The whole file is with 'i' instead of index, should we change each loop?

**Review by SanJSp — DISMISSED** [review]

**Pull request merged**